themebuilder/modules/metadata-generator.js

```diff
@@ -4,66 +4,158 @@ const fs = require("fs");
 
 const themes = require("./themes");
 const getBundleName = require("./bundle-resolver");
+const ModulesHandler = require("../modules/modules-handler");
+
+const stylesDirectory = path.join(__dirname, "..", "..", "styles");
+const themesFileContent = fs.readFileSync(path.join(stylesDirectory, "theme.less"), "utf8");
+const publicWidgets = new ModulesHandler().availableWidgets(themesFileContent).map(w => w.name);
 
 
 const capitalize = (key) => key.charAt(0).toUpperCase() + key.slice(1);
+const getWidgetFromFileName = (fileName) => capitalize(fileName.split("/").pop().split(".")[0]);
+
+const executor = (str, regex, handler) => {
+    let matches;
+    while((matches = regex.exec(str)) !== null) {
+        handler(matches);
+    }
+};
 
 const parseComments = (comments) => {
-    const commentRegex = /@(type|name|typeValues)\s(.+)/g;
     const metaItem = {};
 
-    let matches;
-    while((matches = commentRegex.exec(comments)) !== null) {
+    executor(comments, /@(type|name|typeValues)\s(.+)/g, (matches) => {
         const key = capitalize(matches[1]);
         metaItem[key] = matches[2].trim();
-    }
+    });
 
     return metaItem;
 };
 
 const getMetaItems = (less) => {
-    const commentBlockRegex = /\/\*\*[\n\r]([\s\S]*?)\*\/\s*[\n\r]*([-@a-z_0-9]+):/gim;
     const metaItems = [];
 
-    let matches;
-    while((matches = commentBlockRegex.exec(less)) !== null) {
+    executor(less, /\/\*\*[\n\r]([\s\S]*?)\*\/\s*[\n\r]*([-@a-z_0-9]+):/gim, (matches) => {
         const metaItem = {
             "Key": matches[2]
         };
         metaItems.push(Object.assign(metaItem, parseComments(matches[1])));
-    }
+    });
 
     return metaItems;
 };
 
+const parseImports = (less) => {
+    const imports = [];
+
+    executor(less, /@import(\s+\(.*?\))?\s+['"](.*?\.less)['"]/gi, (matches) => {
+        const fileName = matches[2];
+        const widget = getWidgetFromFileName(fileName);
+        if(!imports.includes(widget)) imports.push(widget);
+    });
+
+    return imports;
+};
+
+const resolveDependencies = (dependencies) => {
+    for(var widget in dependencies) {
+        if(Object.prototype.hasOwnProperty.call(dependencies, widget)) {
+            const widgetDependencies = dependencies[widget];
+
+            for(let i = 0; i < widgetDependencies.length; i++) {
+                const dependencyWidget = widgetDependencies[i];
+                const dependencyList = dependencies[dependencyWidget];
```

Comment:
Shouldn't it be recursive?

Reply:
No. We add the dependency to the end of array and then in any case process it. Theoretically, the algorithm should recursive (it is a tree), but our tree has circle dependencies and recursion can cause the trouble :)